### ISSUE TITLE
Coercing &mut to *const should not create a shared reference

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#56604]
It has long been a rule in Rust that you must not mutate through a shared reference, or a raw pointer obtained from a shared reference.

Unfortunately, that rule currently forbids the following code:
```rust
fn direct_mut_to_const_raw() {
    let x = &mut 0;
    let y: *const i32 = x;
    unsafe { *(y as *mut i32) = 1; }
    assert_eq!(*x, 1);
}
```
The reason for this is that coercing `&mut T` to `*const T` implicitly first creates a shared reference and then coerces that to `*const T`, meaning `y` in the example above is technically a raw pointer obtained from a shared reference.

We should fix our coercion logic to no longer create this intermediate shared reference.

Cc @eddyb @nikomatsakis

**Comment by RalfJung** [comment]
I believe the relevant code might be

https://github.com/rust-lang/rust/blob/f504d3f05985f158c4845816c3d5f7b7a8afeb13/src/librustc_typeck/check/coercion.rs#L734-L752

but I am not sure because this is type *checking*. I have no idea where the code that decides about lowering of coercions to reborrows/casts lives.

**Comment by eddyb** [comment]
> I have no idea where the code that decides about lowering of coercions to reborrows/casts lives.

What do you mean? That code is generating a deref and a borrow, each of those gets turned into the equivalent MIR later, and the result is as if the user wrote `&*` explicitly.

**Comment by eddyb** [comment]
@nikomatsakis I believe the comment above `if is_ref {` has been outdated for quite a while (and makes no sense for NLL), can you confirm?

**Comment by RalfJung** [comment]
@eddyb that code generates "adjustments". I have no idea what that means. I found no use of `AutoBorrow::RawPtr` that would turn this into casts or reborrows or so.

**Comment by eddyb** [comment]
Oh I got confused as to what this is doing, I missed the `AutoBorrow::RawPtr` bit, looks like "borrow" and "cast reference to raw pointer" are fused into one adjustment. The lowering is here, btw:
https://github.com/rust-lang/rust/blob/1c3236afc3560b5781acf942a4746aeb11e5374e/src/librustc_mir/hair/cx/expr.rs#L152-L180

So looking again at the testcase, `y as *mut i32` still goes through the reference->raw pointer coercion logic, but it does so for coercing `&mut T` to `*mut T` which does a mutable reborrow instead of an immutable reborrow (and then there's a separate `*mut T` to `*const T` coercion).

Changing `AutoBorrow::RawPtr(mutbl_b)` to `AutoBorrow::RawPtr(mt_a.mutbl)` would work, except you then *also* need to push `Adjust::MutToConstPointer` to that `vec![...]`, if `mt_a.mutbl != mutbl_b`.

**Comment by RalfJung** [comment]
> Changing AutoBorrow::RawPtr(mutbl_b) to AutoBorrow::RawPtr(mt_a.mutbl) would work, but you then also need to push Adjust::MutToConstPointer to that vec![...], if mt_a.mutbl != mutbl_b.

I did the first but not the last and tests seem to still pass...^^

But yeah, there is a compatibility problem with outstanding shared references. Ouch.

**Comment by eddyb** [comment]
Without `Adjust::MutToConstPointer` the MIR may end up malformed (using `*mut T` where `*const T` is expected), I'm surprised you don't get any errors!

**Comment by RalfJung** [comment]
> using *mut T where *const T is expected

Yeah I figured. run-pass and ui tests all pass, so either the change did nothing or nothing checks the MIR^^

For the backwards compatibility issue:
I think my inclination is that we don't want to reborrow on a cast-to-raw.  That would also make testing Stacked Borrows in miri easier, these implicit unavoidable reborrows are a pain. :P  We'd have to make sure though that the borrow checker understands that after a `(x: &mut T) as *mut T`, `x` is still there -- it hasn't been moved away.

**Comment by eddyb** [comment]
I guess the tricky bit with making it a cast is that what we have to work with is `Operand::{Move,Copy}`.

Instead, we could just add to the MIR a sort of "borrow to raw pointer" (a lot like `AutoBorrow::RawPtr`, really), and then we just need to make sure the old borrowck doesn't treat even a mutable `AutoBorrow::RawPtr` like a real borrow.

**Comment by RalfJung** [comment]
> "borrow to raw pointer"

You mean like https://github.com/rust-lang/rfcs/pull/2582? :D

**Comment by RalfJung** [comment]
> Instead, we could just add to the MIR a sort of "borrow to raw pointer"

Following this idea, I have [updated miri](https://github.com/solson/miri/pull/572) to treat the case of "escaping a ptr to be usable for raw accesses" as just another reborrow: Both trigger a retag; the retag after a cast-to-raw is a bit special because it also retags raw pointers.

So, I am convinced now that the best way forward is to entirely ditch ref-to-raw casts, and encode them all as reborrow-to-raw.

**Comment by RalfJung** [comment]
I found another case of this pattern in the wild, see https://internals.rust-lang.org/t/writing-through-a-pointer-derived-from-a-shared-reference-after-that-reference-is-dead/9078

**Comment by RalfJung** [comment]
This is actually more subtle than I thought. The thing that I did not know when discovering this problem is that the borrow checker treats `*const` and `*mut` different.  The following is safe code:
```rust
let x = &mut 0;
let shared = &*x;
let y = x as *const i32; // if we use *mut here instead, this stops compiling
let _val = *shared;
```
IOW, the borrow checker considers a cast to a `const` raw pointer as just a read access. Also see https://github.com/rust-lang/unsafe-code-guidelines/issues/106 and Cc @matthewjasper 	who brought this to my attention.

This means that if we pull through with the plan to make `x as *const i32` (where `x` is a mutable reference) behave like `x as *mut i32 as *const i32`, that code will stop compiling! Ouch.

So, everybody in https://github.com/rust-lang/unsafe-code-guidelines/issues/106 (including me and @nikomatsakis  but also in particular @SimonSapin) seemed to feel that the following code should be allowed (that's from the OP here):
```rust
// example 1
let x = &mut 0;
let y: *const i32 = x;
unsafe { *(y as *mut i32) = 1; }
```
But how do we all feel about this code?
```rust
// example 2
let x = &mut 0;
let shared = &*x;
let y: *const i32 = x;
let _val = *shared;
unsafe { *(y as *mut i32) = 1; }
```
Here, we cannot possibly maintain the idea that `*const` vs `*mut` does not matter, because the code stops compiling if we use `*mut` instead.

Does that mean that we are fine with example 2 being UB? And if it does, does that change our position about example 1?

## My opinion

Personally I am less sure now about example 1, and feel maybe it should actually be UB, after all. The reason for this is that it leads to the cleanest model, with fewer cases to consider. I will go into several possible models for a bit here.

### Clean and strict

The IMO most clean model (with the usual caveat that we'll not know how clean this ends up being until it got implemented and tested) is that the behavior of cast-to-raw depends *only* on the type of the raw pointer (`mut` vs `const`). We would have a rule that "writing through a raw pointer that was originally created as a `*const` is UB" (except for `UnsafeCell`). This is nice because it would replace/subsume the current rule that "writing through a raw pointer created from a shared reference is UB" (`UnsafeCell` yada yada), since shared references can only be cast to `*const` -- so it would be very consistent in that regard.

Basically, `x as *const T` would be the same as `&*x as *const T`, including all quirks around `UnsafeCell`.

However, this would rule out the example in the OP, and run counter the idea (that I repeated a lot) that `*const` vs `*mut` should not matter.

### Extrapolate from current behavior

Currently, we accept example 1 *if* it gets changed to `let y: *const i32 = x as *mut i32;`. Applying the same rewrite to example 2 does not work, the borrow checker rejects that. Stacked Borrows ignores `*const` vs `*mut` and just checks whether the location is currently frozen; if it is, then the resulting raw pointer cannot be used for writing. If we want to continue along this line, we do have to consider a `&mut` to `*const` cast as a primitive operation instead of being composed of two steps, `&mut` to `*mut` and then `*mut` to `*const`.  So we'd have three reference-to-raw-pointer-casts as three distinct primitive operations:

* `&mut` to `*mut`: Can always be written to (so memory gets unfrozen and a `Raw` tag pushed, in the model).
* `&mut` to `*const`: Can be written to (except `UnsafeCell`) *only if there are no outstanding shared references*. *If* memory is *not* frozen, we push a `Raw` and you can write through this, but if memory is frozen we just allow read access.
* `&` to `*const`: Cannot be written to (except `UnsafeCell`).

This would allow example 1. disallow example 2, and it would also disallow the following:
```rust
// example 2
let x = &mut 0;
let shared = &*x;
let y: *const i32 = x;
unsafe { *(y as *mut i32) = 1; }
```
`shared` gets never used again, but we cannot know that when `y` gets created so we  cannot make `y` a writable pointer. This might seem strange because here, the rewrite to `let y: *const i32 = x as *mut i32;` actually works, but that is because the borrow checker "knows the future" and determines that `shared` does not get used again. We cannot do this in a dynamic semantics such as Stacked Borrows.

## Two-phase

@matthewjasper suggested it might be possible to treat cast-to-`const` as something like a two-phase borrow.  I don't think that makes me happy.^^

**Comment by matthewjasper** [comment]
> @matthewjasper suggested it might be possible to treat cast-to-`const` as something like a two-phase borrow. I don't think that makes me happy.^^

No, I'm was saying that a model that accepts both example 1 and 2 would effectively be modelling two-phase borrows. I can't see any way to allow this without complicating the model significantly.

**Comment by elichai** [comment]
I'd love to see this change.
The current status is too foot gunny IMHO.

I really like the current status of rust that pointers are basically numbers. and they only matter when you dereference, and even then we have no strict alias rules etc. just alignment+size.(and data must be defined which might have implications with padding).

**Comment by RalfJung** [comment]
> I really like the current status of rust that pointers are basically numbers

That is not true at all I am afraid. Even if you are talking about raw pointers only. See [this brief introduction to pointer provenance](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#pointer-provenance). Also, the fact that pointers have provenance is entirely off-topic here.

"Pointers are just numbers" is plain impossible with LLVM as a backend.

> I'd love to see this change.

So I am wondering now what you mean by "this". My assumption of course is that you mean this issue, i.e. specifically casting a mutable reference to a const raw pointer. But then you go on talking about things way broader than that and entirely out of scope for this issue, so I am confused.

Also the reason this issue is open is that we agree it's a foot-gun, what we are looking for is a good solution.

**Comment by elichai** [comment]
> That is not true at all I am afraid. Even if you are talking about raw pointers only. See [this brief introduction to pointer provenance](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#pointer-provenance). Also, the fact that pointers have provenance is entirely off-topic here.

Will read. I heard that a lot so it was more of a quote for me, but interesting to understand. thanks

> Also the reason this issue is open is that we agree it's a foot-gun, what we are looking for is a good solution.

Yes, I mostly wanted to say this because I encountered this UB in the past, mostly in types similar to `NonNull`, that accepted `*const T` instead of accepting `*mut T` and casting. and then you'd do `accepts_const_ptr(&mut something);` without realizing it casts it to a shared reference.

**Comment by RalfJung** [comment]
>  I heard that a lot so it was more of a quote for me

Yeah unfortunately there is a lott of common "wisdom" about pointers in C/C++/Rust out there that is plain wrong. [Pointers are complicated](https://www.ralfj.de/blog/2018/07/24/pointers-and-bytes.html).

> `accepts_const_ptr(&mut something);`

Yeah, that is one if the issues here.
A possible fix is to change desugaring here so that `&mut x as *const _` effectively becomes `&mut x as *mut _ as *const _`, but this might actually break some programs ([see above](https://github.com/rust-lang/rust/issues/56604#issuecomment-477954315), the borrow checker treats cast-to-mutable-raw and cast-to-const-raw differently).

Since the last discussion on this issue, @matthewjasper implemented `&raw`, and that does change some things. For the old example
```rust
let x = &mut 0;
let shared = &*x;
let y = x as *const i32; // if we use *mut here instead, this stops compiling
let _val = *shared;
```
the MIR for the cast is now `_4 = &raw const (*_1);`. There is no longer an explicit shared reference in the MIR. We'll have to test if this actually fixed the bug. It certainly did not change type-checking; the code above still stops compiling when replacing `*const` by `*mut`.

**Comment by RalfJung** [comment]
Even with the changed desugaring, Stacked Borrows still fails on that example because of the way retagging works: `&raw const <expr>` becomes something like
```
_tmp = &raw const <expr>;
Retag([raw] _tmp);
```
Then retagging will consider the type of `_tmp` to see which permissions to acquire for the new pointer, and it'll be `*const T` so it asks for read-only permission. And hence the error.

If we want to make `*const` not significant in terms of UB, we'd have to do something else, like maybe figure out which permissions the source pointer described by `<expr>` has and grant the raw ptr the same permission (but allow it to be shared, i.e., `Unique` becomes `SharedReadWrite`). But that would be a somewhat fundamental change to how Stacked Borrows works for raw pointers. I'll open an issue in the UCG to track this.

**Comment by nikomatsakis** [comment]
@RalfJung 

I have a point of information: Does anyone know what the C (and C++) rules are with respect to `const` pointers/references? In particular, if I 'const cast' away a const reference, am I limited in any way with what kinds of things I can do to the resulting pointer/reference? My understanding is that I am not but this is something I've never felt totally confident about.

I think I would expect raw ponters in Rust to potentially match the rules of C, but not add additional restrictions. Presuming I am correct, then, it seems like this implies that -- for the purposes of stacked borrows -- there are only "raw pointers", and `const` vs `mut` is not relevant. This is certainly the model I have in my head and I feel like it's the one we've expected.

It is true that the borrow checker reports an error in this example:

```rust
let x = &mut 0;
let shared = &*x;
let y = x as *const i32; // if we use *mut here instead, this stops compiling
let _val = *shared;
```

....but it is not obvious to me that it *ought* to do so. In fact, I was surprised, and I expected this any error here to derive from the rules about what sorts of casts and coercions are legal. It seems though that we consider a "cast to mut pointer" to be effectively a kind of *write*, but I don't recall if there were specific motivations behind that (maybe @matthewjasper has a better memory of this).

Looking a bit more into the code, I think this is tied to the `&raw` operations, although when I tested with an older nightly (2019-09-15) I saw the same behavior, so I guess that is preserving some older behavior around casts.

In any case, on master now, looking at example 2:

```rust
// example 2
let x = &mut 0;
let shared = &*x;
let y: *const i32 = x;
let _val = *shared;
unsafe { *(y as *mut i32) = 1; }
```

the `let y: *const i32 = x` effectively becomes a `let y = &raw const *x`. This counts as a **read** of `x`, and therefore is allowed because we effectively have two reads (one from `shared` and one from the `&raw const`).

In contrast, when doing `&raw mut *x`, that is counted as a (instantaneous) **write**, and so we get an error. So really the question is whether `&raw mut` ought to be considered a "write" or a "read", essentially, right? 

With safe references, when you do a `&mut`, we obviously wish to count that as a write, as it enables safe writes to occur forever after. But with `&raw mut`, it doesn't immediately perform any write, and it also does not enable any safe writes to occur in and of itself -- the actual write still requires unsafe -- so I think there's a fine argument for making it a read.

**Comment by elichai** [comment]
> Does anyone know what the C (and C++) rules are with respect to const pointers/references? In particular, if I 'const cast' away a const reference, am I limited in any way with what kinds of things I can do to the resulting pointer/reference? My understanding is that I am not but this is something I've never felt totally confident about.

In C/C++ the const qualifier is a "recommendation" and only matters in variable declaration, ie:
given this function:
```c
void hi (const unsigned char* c) {
*((unsigned char*)c) = 5;
}
```

This is valid:
```c
unsigned char c;
hi(&c);
assert(c == 5);
```
This is UB:
```
const unsigned char c = 42;
hi(&c);
// assert(c == 5); or assert(c == 42);
```

**Comment by RalfJung** [comment]
> With safe references, when you do a &mut, we obviously wish to count that as a write, as it enables safe writes to occur forever after. But with &raw mut, it doesn't immediately perform any write, and it also does not enable any safe writes to occur in and of itself -- the actual write still requires unsafe -- so I think there's a fine argument for making it a read.

This makes sense to me, from the perspective of the borrow checker. However, does that mean that with `x: &T` we also permit `x as *mut _` because that is just a read?

---

If the decision is to adjust borrow checking like that, the next question is how to adjust Stacked Borrows. Currently, the type of the *new* reference/pointer fully determines the permissions of that pointer: when you reborrow to a `*const T`, you get read-only permissions (interior mutability is taken into account); when you reborrow to a `*mut T` you get read-write permissions. If the goal is to make that type distinction not matter, clearly that scheme cannot work.

Instead we could
* check the type of the reference that is cast to a raw pointer to determine the permissions,
* or check the *permissions* of the "parent pointer" to determine new permissions.

The former has some interesting cases when considering `&raw` beyond casts, e.g. `&raw const local.field` *does not have* a reference it "comes from"... so is it always mutable, or do we take mutability of `local` into account? And with `x: & &mut i32`, we can do `&raw const **x`, which while coming from a mutable reference actually comes from an immutable "context", so what about that (this relates to the first question about the borrow checker changes)?

**Comment by nikomatsakis** [comment]
@RalfJung 

> This makes sense to me, from the perspective of the borrow checker. However, does that mean that with `x: &T` we also permit `x as *mut _` because that is just a read?

That is what I am proposing, yes. Now, it would (likely?) be UB to actually do mutation through `*x` -- but it may be that you are just going to be passing `x` to some C function that you know will only do reads (despite not having a `const` pointer type), etc.

Regarding the potential changes to stacked borrows, I guess I have to go refresh my knowledge of the current state of things (or the old state, as the case may be). 

I think the question as to whether a `let x` should be immutable by unsafe code is an interesting one. As you know, I've long felt that `mut` on local variables is more of a "useful heuristic" than anything else -- but it seems like, given that we have this heuristic, we can leverage it to get ourselves more optimization (i.e., because we know that `&raw const x` is still not permitted to do any mutations), and I expect we ought to do so. 

I wonder if the right primitive for `&raw`, from the POV of MIR, is that it carries with it the kinds of permissions it is requesting (which are derived during lowering from the type of the thing to be borrowed) as well as the target type of the pointer (which is independent, oddly enough)?

**Comment by RalfJung** [comment]
> I wonder if the right primitive for &raw, from the POV of MIR, is that it carries with it the kinds of permissions it is requesting (which are derived during lowering from the type of the thing to be borrowed) as well as the target type of the pointer (which is independent, oddly enough)?

If `let` vs `let mut` makes a difference, it's not just the *type* of the thing being borrowed, but also some amount of "metadata" about it. That's why I am not a huge fan of adding this into the mix -- it seems rather messy, TBH.

**Comment by nikomatsakis** [comment]
@RalfJung Indeed, I think it's not really about the type of the thing being borrowed at all, but about the place that it's being borrowed from.

> That's why I am not a huge fan of adding this into the mix -- it seems rather messy, TBH.

I'm not really clear on what "this" refers to here -- what precisely are you not a fan of?

**Comment by RalfJung** [comment]
I'm not a fan of giving the `mut` in `let mut` any operational meaning. Places, like values, should be described by their type alone. (To be fair, that is already not entirely the case -- we also need to track expected alignment, to properly handle `var_of_packed_type.field`.)

Both that `mut` and the `*const` vs `*mut` distinction are things that some people (myself, or at least my past self, included) consider mere syntactic sugar. If we have to give up one of them -- if we have to make one of them actually mean something -- then I'd rather make types mean something than "other annotations on places".

**Comment by nikomatsakis** [comment]
Hmm. 

So I believe what you are saying is this:

* `let mut` should not be meaningful to stacked borrows
* however, by careful construction, we can guarantee that one cannot get a `*mut` reference to a `let` variable *except* by going through a `&` reference, which also ensures that writes through the reference are UB
* therefore, the `let` vs `let mut` distinction cannot be violated by unsafe code, even though it is not "well known" to stacked borrows

And this sort of strikes you as the best of both worlds. =)

It seems to me that it may be possible to make similar guarantees another way. Certainly we could make it an error to do `raw_ref_mut` on a `let` variable (or the deref of a `&T` reference, etc) *without* considering it a "write". 

But I suppose one could do something like this:

```rust
let x = 22;
let p: *const i32 = raw_ref!(x);
let q: *mut i32 = p as *mut i32;
*q += 1;
```

and we have to decide if that is UB. I'm probably out of date in my terminology, but I think that under the "type-based" proposal, the `raw_ref!` would be somehow pushing a "raw, const" onto the stack, and hence the write through `*q` would be UB? But if we lost that type-based approach, then there would be nothing to make the write through `*q` UB except for considering whether `let x` is meaningful or not?

Even though I too have considered `let` vs `let mut` a lint, I think my intuition is the opposite of yours: I would rather make that meaningful than make `*const` vs `*mut` meaningful. I think that is because `let` vs `let mut` feels like the domain of "safe code", and hence should work as expected, and in general I think it'd be useful to the compiler  to be able to be trivially sure `let x` is immutable (though I think that latter point is true in either proposal, and indeed something that the compile can derive simply by looking for how it is borrowed or referenced).

I'm wondering how intrusive of a change is to stacked borrows to have `let` carry a "mutability", it seems like it would basically be a sort of "initial permission" on the stack and nothing more? I suppose at least one complication is that destructors take `&mut self`. =)

**Comment by RalfJung** [comment]
> however, by careful construction, we can guarantee that one cannot get a *mut reference to a let variable except by going through a & reference, which also ensures that writes through the reference are UB

Good point!

> And this sort of strikes you as the best of both worlds. =)

Yes I think you expressed better than I could what I like about the current situation (as counter-intuitive as it is that the type of the raw pointer matters *on the initial ref-to-raw cast*).

> Certainly we could make it an error to do raw_ref_mut on a let variable (or the deref of a &T reference, etc) without considering it a "write".

Fair. So borrowck would still distinguish const and mut raw pointers in the sense that creating a mut raw pointer to an immutable place is disallowed -- but the effect of both, in case they *are* allowed, is the same (unlike now).

> and we have to decide if that is UB. I'm probably out of date in my terminology, but I think that under the "type-based" proposal, the raw_ref! would be somehow pushing a "raw, const" onto the stack, and hence the write through *q would be UB? But if we lost that type-based approach, then there would be nothing to make the write through *q UB except for considering whether let x is meaningful or not?

Hm, very interesting example.  What you call "type-based" is what would happen currently, yes.

>  I suppose at least one complication is that destructors take &mut self. =)

Oh, I hadn't even considered this. That would make it unsound for the compiler to assume that a `let x;` is never mutated if `x` has a destructor!

**Comment by RalfJung** [comment]
> I'm wondering how intrusive of a change is to stacked borrows to have let carry a "mutability", it seems like it would basically be a sort of "initial permission" on the stack and nothing more? 

Well, for once I wouldn't even know how to implement this. Is this information even available in the MIR?

**Comment by nikomatsakis** [comment]
Yes, `let mut` is available in the MIR in the [`mutability` field of the `LocalDecl`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.LocalDecl.html#structfield.mutability). 

I think the key question for us to answer though is what we think about this unsafe code:

```rust
let x = 22;
let p: *const i32 = raw_ref!(x);
let q: *mut i32 = p as *mut i32;
*q += 1;
```

and I suppose the example gets a bit more interesting when you pull the write into a function:

```rust
fn tricky_increment(p: *const i32) {
    *(p as *mut i32) += 1;
}

let x = 22;
tricky_increment(raw_ref!(x));
```

The above indicates an example of "lost optimization" potential I guess.

**Comment by nikomatsakis** [comment]
> Oh, I hadn't even considered this. That would make it unsound for the compiler to assume that a let x; is never mutated if x has a destructor!

I suppose that `DROP(x)` could have some special effect of making the local variable mutable.

**Comment by nikomatsakis** [comment]
So, one thing I am particularly worried about is that we have used (and encouraged) people to use `*const T` as "a owned raw pointer" for some time now (e.g., via `Unique`). It seems like if `*const` *also* means that the data is mutable, we have to find another idiom for that, right? Am I right that this idiom would be invalidated and would require some other support, or else maybe have some weird edge cases associated with it?

**Comment by RalfJung** [comment]
> The above indicates an example of "lost optimization" potential I guess.

This is basically taking the view that a non-`mut` binding should have the same guarantees/optimization potential as shared references. I can see that, though the special case for dropping makes this rather hacky. In Stacked Borrows we'd have to still make the "bottomost" tag read-write, but then store that somewhere on the side and use a different, read-only tag for direct accesses to the local...

> It seems like if *const also means that the data is mutable, we have to find another idiom for that, right?

Did you mean "immutable"?

> Am I right that this idiom would be invalidated and would require some other support, or else maybe have some weird edge cases associated with it?

Invalidated by which proposal? In my proposal, as long as all *public* methods use `*mut T` (like they do for `NonNull`), it doesn't really matter which type is used internally. `raw_ref_mut!(x) as *const _` is still a mutable pointer. It's not the *type* of the raw ptr that matter, it's that we have two different reference-to-raw operations -- and we select which one we used based on the type *at that operation*, but immediately after the cast, the type loses its significance.

**Comment by nikomatsakis** [comment]
@RalfJung 

> It's not the type of the raw ptr that matter, it's that we have two different reference-to-raw operations -- and we select which one we used based on the type at that operation, but immediately after the cast, the type loses its significance.

I see. That makes sense and I think was missing this point somewhat. As far as "which proposal", well, I feel like there are a few sort of semi-proposals on the table. It feels like it'd be nice to have a hackmd or other document that tries to collect the alternatives and summarizes their implications.

That said, I feel like saying that the *initial form of borrow* is meaningful (but not the type itself) might work out reasonably well.

**Comment by RalfJung** [comment]
> I see. That makes sense and I think was missing this point somewhat. As far as "which proposal", well, I feel like there are a few sort of semi-proposals on the table. It feels like it'd be nice to have a hackmd or other document that tries to collect the alternatives and summarizes their implications.

Yeah, we should probably have two or three concrete proposals written down in as much detail as we can somewhere, so we have concrete things to look at. If you point me towards a hackmd I can do that for "whatever Miri currently implements"... need to find a better name for it though. "raw-ptr-type-matters-at-cast-only"?^^

**Comment by nikomatsakis** [comment]
@RalfJung [here is a hackmd](https://hackmd.io/wpBesJ8PSRG0dX4LHcUZUQ), I'll try to populate it some to get started but we should try to fill it out.

**Comment by elichai** [comment]
I don't want to derail the conversation, so if it's unrelated please hide this comment(or I'll delete it).
But if we allow "mutating `let x`" then there's no need for `UnsafeCell`?
(even if that rule is only for pointers obtained by `raw_ref` and *not* by shared reference, we could reimplement `UnsafeCell` with `raw_ref` without the need for a lang item)

**Comment by nikomatsakis** [comment]
Having given this more thought I do feel like we want to rule out the idea of using raw pointers to mutate a `let x` variable. I feel relatively ok about having the *kind of raw borrow* have meaning (much better than I feel about having `*const` vs `*mut` have meaning).

**Comment by RalfJung** [comment]
@nikomatsakis I corrected and expanded the hackmd a bit. Note that right now, I think *both* proposals in there disallow [this code](https://github.com/rust-lang/rust/issues/56604#issuecomment-637656381). We should try to come up with an example that distinguishes the two proposals.

Also, note that `&raw mut *x` does *not* count as a "write access" in current Stacked Borrows. It creates a pointer with `SharedRW` permission, and those pointers are so weak that (a) the write access wouldn't actually get us anything, and (b) in some cases when `SharedRW` is used for interior mutability, that write access would break `RefCell`.

**Comment by RalfJung** [comment]
In terms of distinguishing example, this might be one:
```rust
let x = 22;
let p: *mut i32 = &raw mut x;
*q += 1;
```
"meaningful let" says this is UB, while "meaningful borrow mode" rejects this at compile-time. That seems like a win for "meaningful borrow mode" to me.

Another one:
```rust
let x = &mut 22;
let p: *const i32 = x;
*(p as *mut i32) += 1;
```
Here, "meaningful let" is fine (I think) but "meaningful borrow mode" says this is UB. This is something that could be tweaked [as described here](https://hackmd.io/wpBesJ8PSRG0dX4LHcUZUQ#Mutable-ref-to-raw-behavior).

(I am making some guesses here about your `raw_ref!` syntax, I don't know what you actually had in mind there for "meaningful let". I assumed it is the same as `&raw const`.)

**Comment by nikomatsakis** [comment]
@RalfJung distinguishing examples are good, yes. 

> That seems like a win for "meaningful borrow mode" to me.

Agreed, though of course we could also reject that statically in either proposal if we wished.

This example is an interesting one:

```rust
let x = &mut 22;
let p: *const i32 = x; // meaningful borrow mode: this desugars to `&raw const *x`
*(p as *mut i32) += 1;
```

I think I would indeed expect it to compile, or at least it seems like a footgun (and sort of "busy work") to require the programmer to write `let p: *const i32 = &raw mut *x;`. 

It seems like we're accumulating some "variants" of the core proposals, such as a "meaningful let" that prevents `&raw mut` against a "non-mutable place", or a "meaningful borrow" that takes the source type into account when it desugars coercions.

I guess that having "meaningful let" reject `&raw mut` is a bit "over-agressive" in that there are non-UB programs that would be rejected (e.g., those that never mutate through the resulting reference).

**Comment by RalfJung** [comment]
> I think I would indeed expect it to compile, or at least it seems like a footgun (and sort of "busy work") to require the programmer to write let p: *const i32 = &raw mut *x;.

Incidentally, it is exactly this case which triggered the original creation of this issue. ;)

Here's another distinguishing example, where "meaningful borrow" firmly says "this is UB" and I wonder what your thoughts on that are:
```rust
let x = &mut 22;
let p: *const i32 = &raw const *x; // nothing to desugar
*(p as *mut i32) += 1;
```
We could lint against a const raw borrow of a mutable path. Not sure if we should.

**Comment by nikomatsakis** [comment]
Yeah, that's a good example too, it's not very clear to me just how much difference the explicit operation makes. But at minimum I think it's pretty 'teachable' to say that "the form of the initial borrow determines what you can do with the resulting point later, no matter what type it winds up with" (and the compiler tries to insert the most permissive borrows it can, when the borrow is inserted as a result of a coercion). In other words, I don't know what people will expect absent being given any rules, but I do think that this rule makes a measure of sense, and it fits with the behavior of `&T` and `&mut T`, it seems to me (i.e., the kind of borrow you do is a meaningful operation, whether it be safe or raw).

So I guess this kind of comes back to that more central question of 'what is the criteria we are using to judge whether a rule is intuitive exactly'. I remember @cramertj's input being somewhat surprising, in that he favored *simple, teachable* rules over clever rules that people wouldn't be able to understand (at least that's how I remember it).

**Comment by RalfJung** [comment]
> (and the compiler tries to insert the most permissive borrows it can, when the borrow is inserted as a result of a coercion)

One surprising consequence of this is that `x as *const _` and `&raw const *x` are not the same. Or maybe that's just surprising to me?

**Comment by nikomatsakis** [comment]
I agree that is somewhat surprising.

**Comment by RalfJung** [comment]
@chorman0773 noticed that `&mut [0i32] as *const i32` actually generates a `&raw mut`, while `&mut 0i32 as *const i32` generates a `&raw const` leading to this issue -- that seems odd? I have no idea what happens here during MIR building though.

**Comment by joshlf** [comment]
I've suggested adding a lint for this: https://github.com/rust-lang/rust-clippy/issues/12791

**Comment by briansmith** [comment]
> ```rust
> fn direct_mut_to_const_raw() {
>     let x = &mut 0;
>     let y: *const i32 = x;
>     unsafe { *(y as *mut i32) = 1; }
>     assert_eq!(*x, 1);
> }
> ```
> The reason for this is that coercing &mut T to *const T implicitly first creates a shared reference and then coerces that to > *const T, meaning y in the example above is technically a raw pointer obtained from a shared reference.
> 
> We should fix our coercion logic to no longer create this intermediate shared reference.

`ptr::from_ref(r)` is documented as "This is equivalent to `r as *const T`" and `ptr::from_ref` takes a shared reference as its parameter. Thus we are free to write any `r as *const T` as `ptr::from_ref(r)`. When `r` is a `&mut T`, this will coerce `r` to a shared reference. Consequently, `r as *const T` most also coerce `r` to a shared reference. Otherwise, `ptr::from_ref` would not be equivalent to it.

**Comment by RalfJung** [comment]
> ptr::from_ref(r) is documented as "This is equivalent to r as *const T" and ptr::from_ref takes a shared reference as its parameter. Thus we are free to write any r as *const T as ptr::from_ref(r). 

This argument holds only if r has argument type &T.

**Comment by briansmith** [comment]
> > ptr::from_ref(r) is documented as "This is equivalent to r as *const T" and ptr::from_ref takes a shared reference as its parameter. Thus we are free to write any r as *const T as ptr::from_ref(r). 
> This argument holds only if r has argument type &T.

There are various ways of reading the documentation. One way is to assume that `r` is qualitied by the `r: &T` argument in the signature of the function. The other is to read the statement independently. Not everybody is going to read it the same way as you. And, more importantly, people have already been refactoring code based on the other reading.

**Comment by RalfJung** [comment]
Given that the statement contains `r`, I didn't see any other possible interpretation than to take the domain of quantification of `r` from the type signature. There are obvious counterexamples otherwise: for `r: *mut T`, one of the options does not even compile, so clearly they are not equivalent.

Looks like we should clarify the docs then.

**Comment by briansmith** [comment]
> The reason for this is that coercing `&mut T` to `*const T` implicitly first creates a shared reference and then coerces that to `*const T`, meaning y in the example above is technically a raw pointer obtained from a shared reference.

As I mentioned in #12587, there are at multiple interpretations of `r as *const T` for `r: &mut T`:
1. `((`r as *mut T`) as *const T)` using the transitive rule with rule `&mut T to *mut T` and then rule `*mut T to *const T`.
2. `((r as &T) as *const T` using transitive rule with rule `&mut T to &T` and then rule `&T to *const T`.
3. `Deref::deref(r) as *const T` using transitive rule with the rule for `Deref` and then the rule `&T to *const T`

Maybe the solution here is to precisely define the precedence of the rules.

**Comment by RalfJung** [comment]
That is indeed exactly the open question that is tracked here.

**Comment by RalfJung** [comment]
The code always compiled without errors. Did you even read the issue description? Compile-testing this makes no sense.

**Comment by RodBurman** [comment]
Ralf
  Yes my fault should have dug a bit (lot!) deeper, I have deleted the “poor quality” comment

Thanks for the correction

Rod
